benchmarks/upload_download.py

```diff
@@ -0,0 +1,125 @@
+time to measure something more meaningful.
+"""
+
+from time import time, process_time
+from contextlib import contextmanager
+from tempfile import mkdtemp
+import os
+
+from twisted.trial.unittest import TestCase
+from twisted.internet.defer import gatherResults
+
+from allmydata.util.deferredutil import async_to_deferred
+from allmydata.util.consumer import MemoryConsumer
+from allmydata.test.common_system import SystemTestMixin
+from allmydata.immutable.upload import Data as UData
+from allmydata.mutable.publish import MutableData
+
+
+@contextmanager
+def timeit(name):
+    start = time()
+    start_cpu = process_time()
+    try:
+        yield
+    finally:
+        print(
+            f"{name}: {time() - start:.3f} elapsed, {process_time() - start_cpu:.3f} CPU"
+        )
+
+
+class ImmutableBenchmarks(SystemTestMixin, TestCase):
+    """Benchmarks for immutables."""
+
+    # To use Foolscap, change to True:
+    FORCE_FOOLSCAP_FOR_STORAGE = False
+
+    @async_to_deferred
+    async def setUp(self):
+        SystemTestMixin.setUp(self)
+        self.basedir = os.path.join(mkdtemp(), "nodes")
+
+        # 2 nodes
+        await self.set_up_nodes(2)
+
+        # 1 share
+        for c in self.clients:
+            c.encoding_params["k"] = 1
+            c.encoding_params["happy"] = 1
+            c.encoding_params["n"] = 1
+
+        print()
+
+    @async_to_deferred
+    async def test_upload_and_download_immutable(self):
+        # To test larger files, change this:
+        DATA = b"Some data to upload\n" * 10
+
+        for i in range(5):
+            # 1. Upload:
+            with timeit("  upload"):
+                uploader = self.clients[0].getServiceNamed("uploader")
+                results = await uploader.upload(UData(DATA, convergence=None))
+
+            # 2. Download:
+            with timeit("download"):
+                uri = results.get_uri()
+                node = self.clients[1].create_node_from_uri(uri)
+                mc = await node.read(MemoryConsumer(), 0, None)
+                self.assertEqual(b"".join(mc.chunks), DATA)
+
+    @async_to_deferred
+    async def test_upload_and_download_mutable(self):
+        # To test larger files, change this:
+        DATA = b"Some data to upload\n" * 10
+
+        for i in range(5):
+            # 1. Upload:
+            with timeit("  upload"):
+                result = await self.clients[0].create_mutable_file(MutableData(DATA))
+
+            # 2. Download:
+            with timeit("download"):
+                data = await result.download_best_version()
+                self.assertEqual(data, DATA)
+
+    @async_to_deferred
+    async def test_upload_mutable_in_parallel(self):
+        # To test larger files, change this:
+        DATA = b"Some data to upload\n" * 1_000_000
+        with timeit("  upload"):
+            await gatherResults([
+                self.clients[0].create_mutable_file(MutableData(DATA))
+                for _ in range(20)
+            ])
```

Comment:
On an AMD Ryzen 9 5900HS with 8 physical cores and hyperthreading, the results I see for this benchmark on master@0c507dc5d vs HEAD of the branch are:

* upload: 7.120 elapsed, 7.083 CPU (master)
* upload: 6.796 elapsed, 6.736 CPU (branch)

which looks promising but I was not confident this was a real difference and not the result of noise on the system.  I increased the number of mutables by 10x (with a for loop around the `await ...`) and got

* upload: 68.354 elapsed, 68.127 CPU (master)
* upload: 71.374 elapsed, 71.127 CPU (branch)

:shrug: 

How does this compare to what you see?  Is my system just too noisy, or ... ?

Reply:
The benchmarks aYou want to disable turboboost or whatever AMD calls it when benchmarking.

With no hyperthreading (decided it's pointless or actively bad for my use cases), and turboboost disabled, I did four runs each of the new parallelism benchmark:

#### With running in thread

  upload: 5.433 elapsed, 6.086 CPU
  upload: 6.656 elapsed, 7.247 CPU
  upload: 6.218 elapsed, 6.913 CPU
  upload: 5.944 elapsed, 6.561 CPU

#### With running in thread disabled

  upload: 6.842 elapsed, 6.842 CPU
  upload: 6.831 elapsed, 6.823 CPU
  upload: 6.629 elapsed, 6.629 CPU
  upload: 7.431 elapsed, 7.430 CPU

Reply:
This benchmark doesn't complete for me anymore :/

```
    @async_to_deferred                                                                   
    async def test_upload_mutable_in_parallel(self):                                                                                                                              
        # To test larger files, change this:                                             
        DATA = b"Some data to upload\n" * 1_000_000                                                                                                                               
        with timeit("  upload"):                                                                                                                                                  
>           await gatherResults([                                                                                                                                                 
                self.clients[0].create_mutable_file(MutableData(DATA))                                                                                                            
                for _ in range(20)                                                                                                                                                
            ])                                                                                                                                                                    
E           twisted.internet.defer.FirstError: FirstError[#0, [Failure instance: Traceback (failure with no frames): <class 'allmydata.mutable.common.NotEnoughServersError'>: ("P
ublish ran out of good servers, last failure was: [Failure instance: Traceback: <class 'twisted.internet.error.ConnectingCancelledError'>: HostnameAddress(hostname=b'127.0.0.1', 
port=53659)\n/home/exarkun/VirtualEnvs/tahoe-lafs/lib/python3.9/site-packages/twisted/internet/defer.py:735:errback\n/home/exarkun/VirtualEnvs/tahoe-lafs/lib/python3.9/site-packa
ges/twisted/internet/defer.py:798:_startRunCallbacks\n/home/exarkun/VirtualEnvs/tahoe-lafs/lib/python3.9/site-packages/twisted/internet/defer.py:892:_runCallbacks\n/home/exarkun/
VirtualEnvs/tahoe-lafs/lib/python3.9/site-packages/twisted/internet/defer.py:1792:gotResult\n--- <exception caught here> ---\n/home/exarkun/VirtualEnvs/tahoe-lafs/lib/python3.9/s
ite-packages/twisted/internet/defer.py:1693:_inlineCallbacks\n/home/exarkun/VirtualEnvs/tahoe-lafs/lib/python3.9/site-packages/twisted/python/failure.py:518:throwExceptionIntoGen
erator\n/home/exarkun/VirtualEnvs/tahoe-lafs/lib/python3.9/site-packages/allmydata/storage_client.py:1445:slot_testv_and_readv_and_writev\n/home/exarkun/VirtualEnvs/tahoe-lafs/li
b/python3.9/site-packages/twisted/internet/defer.py:1693:_inlineCallbacks\n/home/exarkun/VirtualEnvs/tahoe-lafs/lib/python3.9/site-packages/twisted/python/failure.py:518:throwExc
eptionIntoGenerator\n/home/exarkun/VirtualEnvs/tahoe-lafs/lib/python3.9/site-packages/allmydata/storage/http_client.py:854:read_test_write_chunks\n]", None)                      
E           ]]                                                                                                                                                                   
```

Reply:
Fixed I hope.